app/src/main/java/org/mozilla/fenix/utils/OnVisualCompletenessExecutionQueue.kt

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.utils
+
+import android.view.View
+import kotlinx.coroutines.Job
+import mozilla.components.support.base.log.Log
+
+/**
+ * A queue of paused `Job`s whose execution is resumed when Fenix is visually complete.
+ *
+ * This queue contains paused `Job`s. Each Job in the queue will be restarted once Fenix
+ * is visually complete (as long as it has not be cancelled since it was added to the queue).
+ * Each Job will execute on the thread on which it was originally scheduled.
+ *
+ * This is threadsafe.
+ */
+object OnVisualCompletenessExecutionQueue {
+    internal var jobs = mutableListOf<Job>()
+    val condition = Object()
+    private const val visualCompletenessExecutionQueueLogTag = "OnVisualCompletenessExecutionQueue"
+
+    /**
+     * Add a paused Job to the queue. If the job is completed or active, it will *not* be added
+     * to the queue.
+     *
+     * @param j: The paused Job to add to the queue.
+     */
+    @Synchronized
+    fun add(j: Job) {
+        if (j.isActive || j.isCompleted) {
+            Log.log(
+                Log.Priority.WARN,
+                visualCompletenessExecutionQueueLogTag,
+                null,
+                "Not adding a Job that is active or complete."
+            )
+            throw VisualCompletenessExecutionQueueException("Job is active or complete.")
+        }
+
+        jobs.add(j)
+    }
+
+    /**
+     * Posts a runnable action to the view and triggers the queue to start executing its jobs.
+     *
+     *@param v: A view that maintains its message queue.
+     */
+    @Synchronized
+    fun post(v: View) {
+        v.post {
```

Comment:
Does this actually run after visual completeness? It seems like it will run after the first frame is drawn and the framework yields the UI thread to our posted Runnable, i.e. before the tabs are restored.